add etcd test analyzer build and integrate into measure-test-flakiness workflow

chaochn47 opened the pull request:
The PR is to address the ask from https://github.com/etcd-io/etcd/issues/13167#issuecomment-1474791002. @serathius

The workflow stays running every Sunday midnight. 

Measure-test-flakiness workflow output will additionally show the top contributor of the failed Tests name in Test/Test-arm workflow (only covers integration tests right now) in the past 1 week.

Example output https://github.com/chaochn47/etcd/actions/runs/4546435089/jobs/8015061879

---

Running against E2E workflows does not produce meaningful output, need to investigate more in the tool
```
dev-dsk-chaochn-2c-a26acd76 % bin/etcd-test-analyzer run -token $GITHUB_TOKEN -max-age=168h -workflow E2E -branch main
Fetching workflow runs from GitHub
Analyzing workflow "E2E"
    [E2E]: Finding workflow runs
    [E2E]: Downloading artifacts
    [E2E]: Downloading artifacts for E2E to "/tmp/test-results-2184436320"
    ...
Runs: 0, Pass: 0, Fail: 0, Pcnt: NaN
Analyzing workflow "E2E-arm64"
    [E2E-arm64]: Finding workflow runs
    [E2E-arm64]: Downloading artifacts
    [E2E-arm64]: Downloading artifacts for E2E-arm64 to "/tmp/test-results-2423396168"
    [E2E-arm64]: no artifacts found
    ...
```

Maybe @endocrimes can help shed lights on the issue. 

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

serathius commented:
> Running against E2E workflows does not produce meaningful output, need to investigate more in the tool

E2e tests don't upload Junit reports. https://github.com/etcd-io/etcd/pull/13152
Definetly something worth doing now, with the new tool.

chaochn47 commented:
> E2e tests don't upload Junit reports. https://github.com/etcd-io/etcd/pull/13152

Appreciate the insights!! I can take a look into this.

chaochn47 commented on scripts/build_lib.sh:
```diff
@@ -4,6 +4,10 @@ source ./scripts/test_lib.sh
 
 GIT_SHA=$(git rev-parse --short HEAD || echo "GitNotFound")
 VERSION_SYMBOL="${ROOT_MODULE}/api/v3/version.GitSHA"
+GO_LDFLAGS="${GO_LDFLAGS:-"-s -w"}"
+GOOS="${GOOS:-"linux"}"
+GOARCH="${GOARCH:-"amd64"}"
+BINDIR="${BINDIR:-"bin"}"
```
Requires default value to pass unbound variable shell check.

serathius replied:
This is inconsistent with proposal in https://github.com/etcd-io/etcd/pull/15519
cc @fuweid 
Could you align on this to avoid conflicts?

chaochn47 replied:
Yeah, I will rebase the commit on the referenced PR.

serathius commented on .github/workflows/measure-test-flakiness.yaml:
```diff
@@ -12,6 +12,9 @@ jobs:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      run: |
+        ./scripts/measure-test-flakiness.sh
+        GOARCH=amd64 ./scripts/build_etcd_test_analyzer.sh
```
```suggestion
        make bin/etcd-test-analyser
```
If we are adding Makefile target for local use, let's use it in workflow. This way we test Makefile instead of letting it rot.

chaochn47 replied:
done. Thanks for suggestion.

serathius commented on scripts/build_etcd_test_analyzer.sh:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+# http://redsymbol.net/articles/unofficial-bash-strict-mode/
```
```suggestion
```

Not needed to comment it, we are adding script bash everywhere https://github.com/etcd-io/etcd/pull/15519.

serathius commented on Makefile:
```diff
@@ -9,6 +9,13 @@ build:
 
+TEMP_TEST_ANALYZER_DIR=/tmp/etcd-test-analyzer
+bin/etcd-test-analyzer: $(TEMP_TEST_ANALYZER_DIR)/*
+	GO_BUILD_FLAGS="${GO_BUILD_FLAGS} -v" ./scripts/build_etcd_test_analyzer.sh
```
Not super convinced that we should put it as a separate script. Building an it seems simple enough that it can be put inline in makefile.

serathius commented on Makefile:
```diff
@@ -9,6 +9,13 @@ build:
+	GO_BUILD_FLAGS="${GO_BUILD_FLAGS} -v" ./scripts/build_etcd_test_analyzer.sh
+
+$(TEMP_TEST_ANALYZER_DIR)/*:
+	pushd /tmp; git clone "https://github.com/endocrimes/etcd-test-analyzer.git"; popd
```
```suggestion
	cd /tmp; git clone "https://github.com/endocrimes/etcd-test-analyzer.git"
```

fuweid commented on Makefile:
```diff
@@ -9,6 +9,14 @@ build:
+TEMP_TEST_ANALYZER_DIR=/tmp/etcd-test-analyzer
+TEST_ANALYZER_BIN=${PWD}
+bin/etcd-test-analyzer: $(TEMP_TEST_ANALYZER_DIR)/*
+	cd ${TEMP_TEST_ANALYZER_DIR}/; go build -v -o ${TEST_ANALYZER_BIN}/bin/etcd-test-analyzer; ${TEST_ANALYZER_BIN}/bin/etcd-test-analyzer -h
```
The project supports use `make`. It can be

```
make -C ${TEMP_TEST_ANALYZER_DIR} build
mkdir -p ./bin
install ${TEMP_TEST_ANALYZER_DIR}/bin/etcd-test-analyzer ./bin/
./bin/etcd-test-analyzer -h
```

chaochn47 replied:
Nice. I would like to follow that project Makefile build target.

fuweid reviewed (approved):
LGTM(non-binding)

Thanks

> Just share some ideas about makefile~

chaochn47 commented on Makefile:
```diff
@@ -9,6 +9,14 @@ build:
+	cd ${TEMP_TEST_ANALYZER_DIR}/; go build -v -o ${TEST_ANALYZER_BIN}/bin/etcd-test-analyzer; ${TEST_ANALYZER_BIN}/bin/etcd-test-analyzer -h
+
+$(TEMP_TEST_ANALYZER_DIR)/*:
+	cd /tmp; git clone "https://github.com/endocrimes/etcd-test-analyzer.git"
```
Thanks. It's good to know. Will update.

chaochn47 commented:
Addressed all the comments. 

The example output in run on push workflow. https://github.com/chaochn47/etcd/actions/runs/4546435089/jobs/8015061879. The second one is tracked in 

```
Runs: 15, Pass: 13, Fail: 2, Pcnt: 0.866667
│───────────────────────────────────│────────────────────────────────────────────────────────│───────────────│─────────────────────────────────────────────────────────│
│ TEST NAME                         │ PACKAGE                                                │ FAILURE COUNT │ RECENT FAILURE                                          │
│───────────────────────────────────│────────────────────────────────────────────────────────│───────────────│─────────────────────────────────────────────────────────│
│ ExampleCluster_memberAddAsLearner │ go.etcd.io/etcd/tests/v3/integration/clientv3/examples │             1 │ https://github.com/etcd-io/etcd/actions/runs/4495063935 │
│ TestV3WatchRestoreSnapshotUnsync  │ go.etcd.io/etcd/tests/v3/integration                   │             1 │ https://github.com/etcd-io/etcd/actions/runs/4516121595 │
│───────────────────────────────────│────────────────────────────────────────────────────────│───────────────│─────────────────────────────────────────────────────────│
```

ahrtr reviewed (approved):
Looks great, thank you @chaochn47
